### PR TITLE
favorite discount

### PR DESCRIPTION
**Pull request by OlegKandalov**

### CODE REVIEW

File: src/main/java/com/exadel/discount/platform/service/FavoriteDiscountService.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    public FavoriteDiscountResponseDto like(UUID discountId) {
+        Discount discount = Optional.of(
+                discountRepository.findDiscountByIdAndIsDeletedAndEndDateAfter(discountId, false, ZonedDateTime.now())
+        ).orElseThrow(() -> new NotFoundException("Discount with id " + discountId + " doesn't exist or was finished"));
+
+        MyUserDetails details = (MyUserDetails) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
+
+        boolean likedDiscount = favoriteDiscountRepository.existsByDiscountIdAndUserId(discountId, details.getUserId());
+
+        if (!likedDiscount) {
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/service/FavoriteDiscountService.java:38]: if exists -> throw. Not if not exists -> do something

File: src/main/java/com/exadel/discount/platform/web/FavoriteDiscountController.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.UUID;
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/discount/favorite/{id}")
```
Comment (ohalushchak-exadel) [inline src/main/java/com/exadel/discount/platform/web/FavoriteDiscountController.java:16]: /discount/{id}/favorite

File: src/main/resources/liquibase/20.07.2021_oleg_create_favorite_discount.xml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        <createTable tableName="favorite_discount">
+            <column defaultValueComputed="uuid_generate_v4()" name="fd_id" type="UUID">
+                <constraints nullable="false" primaryKey="true" primaryKeyName="pk_favorite_discount"/>
+            </column>
+            <column name="fd_discount_id" type="UUID">
+                <constraints nullable="false"/>
+            </column>
+            <column name="fd_user_id" type="UUID">
+                <constraints nullable="false"/>
+            </column>
```
Comment (ohalushchak-exadel) [inline src/main/resources/liquibase/20.07.2021_oleg_create_favorite_discount.xml:16]: foreign key

File: src/main/resources/liquibase/20.07.2021_oleg_create_favorite_discount.xml [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+                                 referencedColumnNames="d_id"/>
+    </changeSet>
+
+    <changeSet id="fk_user_favorite_discount" author="oleg">
+        <addForeignKeyConstraint baseTableName="favorite_discount"
+                                 baseColumnNames="fd_user_id"
+                                 constraintName="fk_favorite_discount_fd_user_id"
+                                 referencedTableName="users"
+                                 referencedColumnNames="u_id"/>
+    </changeSet>
```
Comment (ohalushchak-exadel) [inline src/main/resources/liquibase/20.07.2021_oleg_create_favorite_discount.xml:35]: just add FKs in the `constraints` tags of the changeset above

File: src/main/java/com/exadel/discount/platform/model/dto/FavoriteDiscountResponseDto.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package com.exadel.discount.platform.model.dto;
+
+import lombok.Data;
+
+import javax.validation.constraints.NotNull;
+import java.time.ZonedDateTime;
+import java.util.UUID;
+
+@Data
+public class FavoriteDiscountResponseDto {
```
[the remark this answers is not in the document]
Comment (nkharitoniuk) [reply src/main/java/com/exadel/discount/platform/model/dto/FavoriteDiscountResponseDto.java:10]: You mean returun it to front? I think it is not necessary.

**Review by ohalushchak-exadel — APPROVED** [review]

**Review by nkharitoniuk — APPROVED** [review]

**Pull request merged**